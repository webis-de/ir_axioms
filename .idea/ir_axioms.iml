<?xml version="1.0" encoding="UTF-8"?>
<module type="PYTHON_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/venv" />
      <excludeFolder url="file://$MODULE_DIR$/build" />
      <excludeFolder url="file://$MODULE_DIR$/dist" />
      <excludeFolder url="file://$MODULE_DIR$/venv-base/lib" />
      <excludeFolder url="file://$MODULE_DIR$/.pytest_cache" />
      <excludeFolder url="file://$MODULE_DIR$/ir_axioms.egg-info" />
      <excludeFolder url="file://$MODULE_DIR$/venv-base" />
      <excludeFolder url="file://$MODULE_DIR$/tmp" />
      <excludeFolder url="file://$MODULE_DIR$/venv-pyserini" />
      <excludeFolder url="file://$MODULE_DIR$/venv-pyterrier" />
      <excludeFolder url="file://$MODULE_DIR$/data" />
      <excludeFolder url="file://$MODULE_DIR$/venv-test" />
      <excludeFolder url="file://$MODULE_DIR$/.idea/dataSources" />
      <excludeFolder url="file://$MODULE_DIR$/examples/cache" />
      <excludeFolder url="file://$MODULE_DIR$/examples/trec-system-runs" />
      <excludeFolder url="file://$MODULE_DIR$/examples/pyterrier" />
      <excludeFolder url="file://$MODULE_DIR$/examples/data" />
      <excludeFolder url="file://$MODULE_DIR$/examples/results" />
      <excludeFolder url="file://$MODULE_DIR$/venv-3.7" />
      <excludeFolder url="file://$MODULE_DIR$/examples/.ipynb_checkpoints" />
      <excludeFolder url="file://$MODULE_DIR$/examples/msmarco-index" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
  <component name="PackageRequirementsSettings">
    <option name="requirementsPath" value="$MODULE_DIR$/setup.cfg" />
  </component>
</module>