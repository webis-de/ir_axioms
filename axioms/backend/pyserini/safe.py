# Check if Pyserini is installed.
try:
    import pyserini
    assert pyserini
except ImportError as error:
    raise ImportError(
        "The Pyserini backend requires that 'pyserini' is installed."
    ) from error

from pyserini.index import IndexReader as _IndexReader
from pyserini.search import SimpleSearcher as _SimpleSearcher
from jnius import autoclass as _autoclass

# Re-export modules.
IndexReader = _IndexReader
SimpleSearcher = _SimpleSearcher
autoclass = _autoclass
