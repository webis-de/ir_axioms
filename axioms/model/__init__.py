# Re-export from sub-modules.

from axioms.model.base import (  # noqa: F401
    Input,
    Output,
    Preference,
    PreferenceMatrix,
    Mask,
    MaskMatrix,
)

from axioms.model.retrieval import (  # noqa: F401
    Query,
    TextQuery,
    Document,
    TextDocument,
    ScoredDocument,
    ScoredTextDocument,
    RankedDocument,
    JudgedDocument,
    IndexContext,
)

from axioms.model.generation import (  # noqa: F401
    GenerationInput,
    GenerationOutput,
    Aspect,
    Aspects,
)
