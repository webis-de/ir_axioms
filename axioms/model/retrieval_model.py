from dataclasses import dataclass
from typing import Union


@dataclass(frozen=True)
class Tf:
    pass


@dataclass(frozen=True)
class TfIdf:
    pass


@dataclass(frozen=True)
class BM25:
    k_1: float = 1.2
    k_3: float = 8
    b: float = 0.75


@dataclass(frozen=True)
class PL2:
    c: float = 0.1


@dataclass(frozen=True)
class DirichletLM:
    mu: float = 1000


# Aliases
QL = DirichletLM

# Type union
RetrievalModel = Union[Tf, TfIdf, BM25, PL2, DirichletLM]
