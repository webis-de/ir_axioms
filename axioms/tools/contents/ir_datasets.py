from dataclasses import dataclass
from functools import cached_property
from typing import Iterable, Mapping, Union

from ir_datasets import Dataset, load as irds_load
from ir_datasets.formats import GenericDoc, GenericQuery
from ir_datasets.indices import Docstore

from axioms.model.retrieval import Document, Query
from axioms.tools.contents.base import TextContents
from axioms.tools.contents.simple import FallbackTextContentsMixin


@dataclass(frozen=True, kw_only=True)
class IrdsDocumentTextContents(
    FallbackTextContentsMixin[Document], TextContents[Document]
):
    dataset: Union[Dataset, str]

    @cached_property
    def _dataset(self) -> Dataset:
        if isinstance(self.dataset, Dataset):
            return self.dataset
        else:
            return irds_load(self.dataset)

    @cached_property
    def _documents_store(self) -> Docstore:
        return self._dataset.docs_store()

    def contents(self, input: Document) -> str:
        irds_document: GenericDoc = self._documents_store.get(input.id)
        return irds_document.default_text()


@dataclass(frozen=True, kw_only=True)
class IrdsQueryTextContents(FallbackTextContentsMixin[Query], TextContents[Query]):
    dataset: Union[Dataset, str]

    @cached_property
    def _dataset(self) -> Dataset:
        if isinstance(self.dataset, Dataset):
            return self.dataset
        else:
            return irds_load(self.dataset)

    @cached_property
    def _queries(self) -> Mapping[str, GenericQuery]:
        queries: Iterable[GenericQuery] = self._dataset.queries_iter()
        return {query.query_id: query.default_text() for query in queries}

    def contents(self, input: Query) -> str:
        return self._queries[input.id]
