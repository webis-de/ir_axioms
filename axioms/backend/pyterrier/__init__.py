# Check if PyTerrier is installed.
try:
    import pyterrier # noqa: F401
except ImportError as error:
    raise ImportError(
        "The PyTerrier backend requires that 'python-terrier' is installed."
    ) from error

from dataclasses import dataclass
from functools import lru_cache, cached_property
from pathlib import Path
from re import split
from typing import Optional, Union, Callable, NamedTuple, Sequence, TYPE_CHECKING, TypeAlias, Any

from ir_datasets import Dataset, load
from ir_datasets.indices import Docstore
from jnius import autoclass
from pyterrier.datasets import IRDSDataset
from pyterrier.java import started

from axioms.model import Query, Document, TextDocument, IndexContext

# Load Java classes.
if TYPE_CHECKING or not started():
    StringReader: TypeAlias = Any
    Index: TypeAlias = Any
    IndexRef: TypeAlias = Any
    IndexFactory: TypeAlias = Any
    MetaIndex: TypeAlias = Any
    Lexicon: TypeAlias = Any
    CollectionStatistics: TypeAlias = Any
    Tokeniser: TypeAlias = Any
    EnglishTokeniser: TypeAlias = Any
    TermPipelineAccessor: TypeAlias = Any
    BaseTermPipelineAccessor: TypeAlias = Any
    ApplicationSetup: TypeAlias = Any
else:
    StringReader = autoclass("java.io.StringReader")
    Index = autoclass("org.terrier.structures.Index")
    IndexRef = autoclass('org.terrier.querying.IndexRef')
    IndexFactory = autoclass('org.terrier.structures.IndexFactory')
    MetaIndex = autoclass("org.terrier.structures.MetaIndex")
    Lexicon = autoclass("org.terrier.structures.Lexicon")
    CollectionStatistics = autoclass("org.terrier.structures.CollectionStatistics")
    Tokeniser = autoclass("org.terrier.indexing.tokenisation.Tokeniser")
    EnglishTokeniser = autoclass(
        "org.terrier.indexing.tokenisation.EnglishTokeniser"
    )
    TermPipelineAccessor = autoclass("org.terrier.terms.TermPipelineAccessor")
    BaseTermPipelineAccessor = autoclass(
        "org.terrier.terms.BaseTermPipelineAccessor"
    )
    ApplicationSetup = autoclass('org.terrier.utility.ApplicationSetup')

ContentsAccessor = Union[str, Callable[[NamedTuple], str]]


@dataclass(frozen=True, kw_only=True)
class TerrierIndexContext(IndexContext):
    index_location: Union[Index, IndexRef, Path, str]
    dataset: Optional[Union[Dataset, str, IRDSDataset]] = None
    contents_accessor: Optional[ContentsAccessor] = "text"
    tokeniser: Optional[Tokeniser] = None
    cache_dir: Optional[Path] = None

    @cached_property
    def _index(self) -> Index:
        if not TYPE_CHECKING and isinstance(self.index_location, Index):
            return self.index_location
        elif not TYPE_CHECKING and isinstance(self.index_location, IndexRef):
            return IndexFactory.of(self.index_location)
        elif isinstance(self.index_location, str):
            return IndexFactory.of(self.index_location)
        elif isinstance(self.index_location, Path):
            return IndexFactory.of(str(self.index_location.absolute()))
        else:
            raise ValueError(
                f"Cannot load index from location {self.index_location}."
            )

    def __str__(self):
        return f'TerrierIndexContext({str(self.index_location).split("/")[-1].split(" ")[0]})'

    @cached_property
    def _meta_index(self) -> MetaIndex:
        meta_index = self._index.getMetaIndex()
        if meta_index is None:
            raise ValueError(
                f"Index {self.index_location} does not have a metaindex."
            )
        return meta_index

    @cached_property
    def _meta_index_keys(self) -> Sequence[str]:
        return tuple(str(key) for key in self._meta_index.getKeys())

    @cached_property
    def _lexicon(self) -> Lexicon:
        return self._index.getLexicon()

    @cached_property
    def _collection_statistics(self) -> CollectionStatistics:
        return self._index.getCollectionStatistics()

    @cached_property
    def _dataset(self) -> Optional[Dataset]:
        if self.dataset is None:
            return None
        elif isinstance(self.dataset, Dataset):
            return self.dataset
        elif isinstance(self.dataset, str):
            return load(self.dataset)
        elif isinstance(self.dataset, IRDSDataset):
            return self.dataset.irds_ref()
        else:
            raise ValueError(f"Cannot load dataset {self.dataset}.")

    @cached_property
    def _dataset_doc_store(self) -> Optional[Docstore]:
        if self._dataset is None:
            return None
        else:
            return self._dataset.docs_store()

    @cached_property
    def document_count(self) -> int:
        return int(self._collection_statistics.numberOfDocuments)

    @lru_cache(None)
    def document_frequency(self, term: str) -> int:
        entry = self._lexicon.getLexiconEntry(term)
        if entry is None or entry.getNumberOfEntries() == 0:
            del entry
            return 0
        else:
            document_frequency = int(entry.getDocumentFrequency())
            del entry
            return document_frequency

    @lru_cache(None)
    def _document_contents(self, document_id: str) -> str:
        # Shortcut when ir_dataset is specified.
        documents_store = self._dataset_doc_store
        if documents_store is not None:
            try:
                store_document = documents_store.get(document_id)
                if self.contents_accessor is None:
                    return store_document.text
                elif isinstance(self.contents_accessor, str):
                    return getattr(store_document, self.contents_accessor)
                else:
                    return self.contents_accessor(store_document)
            except KeyError:
                # Document not found. Assume empty content.
                return ""

        if (
                self.contents_accessor is None or
                not isinstance(self.contents_accessor, str)
        ):
            raise ValueError(
                f"Cannot load contents "
                f"from metaindex field {self.contents_accessor}."
            )

        if self.contents_accessor not in self._meta_index_keys:
            raise ValueError(
                f"Index {self.index_location} did not have "
                f"requested metaindex key {self.contents_accessor}. "
                f"Keys present in metaindex "
                f"are {self._meta_index_keys}."
            )

        doc_id = int(self._meta_index.getDocument("docno", document_id))
        contents = str(self._meta_index.getItem(
            self.contents_accessor,
            doc_id,
        ))
        return contents

    def document_contents(self, document: Document) -> str:
        # Shortcut when text is given in the document.
        if isinstance(document, TextDocument):
            return document.contents

        return self._document_contents(document.id)

    @cached_property
    def _tokeniser(self) -> Tokeniser:
        if self.tokeniser is None:
            return EnglishTokeniser()
        return self.tokeniser

    @cached_property
    def _term_pipelines(self) -> Sequence[TermPipelineAccessor]:
        term_pipelines = str(ApplicationSetup.getProperty(
            "termpipelines",
            "Stopwords,PorterStemmer"
        ))
        return tuple(
            BaseTermPipelineAccessor(pipeline)
            for pipeline in split(r"\s*,\s*", term_pipelines.strip())
        )

    @lru_cache(None)
    def _terms(self, text: str) -> Sequence[str]:
        reader = StringReader(text)
        terms = tuple(
            str(term)
            for term in self._tokeniser.tokenise(reader)
            if term is not None
        )
        del reader

        for pipeline in self._term_pipelines:
            terms = tuple(
                str(term)
                for term in map(pipeline.pipelineTerm, terms)
                if term is not None
            )
        return terms

    def terms(
            self,
            query_or_document: Union[Query, Document]
    ) -> Sequence[str]:
        text = self.contents(query_or_document)
        return self._terms(text)
