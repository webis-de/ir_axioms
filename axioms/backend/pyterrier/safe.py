# Check if PyTerrier is installed.
try:
    import pyterrier

    assert pyterrier
    # Ensure that the Terrier JVM has started.
    from pyterrier import started

    assert started()
except ImportError as error:
    raise ImportError(
        "The PyTerrier backend requires that 'python-terrier' is installed."
    ) from error

from jnius import autoclass as _autoclass
from pyterrier.apply import generic as _generic
from pyterrier.index import IterDictIndexer as _IterDictIndexer
from pyterrier.datasets import IRDSDataset as _IRDSDataset
from pyterrier.transformer import (
    Transformer as _Transformer, TransformerBase as _TransformerBase,
    IdentityTransformer as _IdentityTransformer,
    EstimatorBase as _EstimatorBase
)

# Re-export modules.
IterDictIndexer = _IterDictIndexer
IRDSDataset = _IRDSDataset
Transformer = _Transformer
TransformerBase = _TransformerBase
IdentityTransformer = _IdentityTransformer
EstimatorBase = _EstimatorBase
autoclass = _autoclass
generic = _generic
