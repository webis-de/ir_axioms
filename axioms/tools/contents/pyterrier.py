from axioms.utils.libraries import is_pyterrier_installed

if is_pyterrier_installed():

    from dataclasses import dataclass
    from functools import lru_cache, cached_property
    from pathlib import Path
    from typing import Union, Sequence, Any

    from pyterrier.java import (
        required as pt_java_required,
        autoclass as pt_java_autoclass,
    )
    from pyterrier.terrier import IndexFactory

    from axioms.model.retrieval import Document
    from axioms.tools.contents.base import TextContents
    from axioms.tools.contents.simple import FallbackTextContentsMixin

    @pt_java_required
    def autoclass(*args, **kwargs) -> Any:
        return pt_java_autoclass(*args, **kwargs)

    Index = autoclass("org.terrier.structures.Index")
    IndexRef = autoclass("org.terrier.querying.IndexRef")
    MetaIndex = autoclass("org.terrier.structures.MetaIndex")

    @pt_java_required
    @dataclass(frozen=True, kw_only=True)
    class TerrierDocumentTextContents(
        FallbackTextContentsMixin[Document], TextContents[Document]
    ):
        index_location: Union[Index, IndexRef, Path, str]  # type: ignore
        text_field: str = "text"

        @cached_property
        def _index(self) -> Any:
            if isinstance(self.index_location, Index):
                return self.index_location
            elif isinstance(self.index_location, IndexRef):
                return IndexFactory.of(self.index_location)
            elif isinstance(self.index_location, str):
                return IndexFactory.of(self.index_location)
            elif isinstance(self.index_location, Path):
                return IndexFactory.of(str(self.index_location.absolute()))
            else:
                raise ValueError(
                    f"Cannot load index from location {self.index_location}."
                )

        @cached_property
        def _meta_index(self) -> Any:
            meta_index = self._index.getMetaIndex()
            if meta_index is None:
                raise ValueError(
                    f"Index {self.index_location} does not have a metaindex."
                )
            return meta_index

        @cached_property
        def _meta_index_keys(self) -> Sequence[str]:
            return [str(key) for key in self._meta_index.getKeys()]

        @lru_cache(None)
        def _document_contents(self, document_id: str) -> str:
            if self.text_field not in self._meta_index_keys:
                raise ValueError(
                    f"Index {self.index_location} did not have "
                    f'requested metaindex key "{self.text_field}". '
                    f"Keys present in metaindex "
                    f"are {self._meta_index_keys}."
                )

            doc_id = int(self._meta_index.getDocument("docno", document_id))
            contents = str(
                self._meta_index.getItem(
                    self.text_field,
                    doc_id,
                )
            )
            return contents

        def contents(self, inputocument) -> str:
            return self._document_contents(input)

else:
    TerrierDocumentTextContents = NotImplemented  # type: ignore
