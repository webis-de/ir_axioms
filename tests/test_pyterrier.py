from pyterrier import started, init

if not started():
    init(tqdm="auto")

from pathlib import Path

from ir_datasets import load, Dataset
from pytest import fixture, approx

from ir_axioms.backend.pyterrier import TerrierIndexContext
from ir_axioms.backend.pyterrier.safe import IterDictIndexer
from ir_axioms.model import Query, Document
from ir_axioms.model.context import IndexContext
from ir_axioms.model.retrieval_model import Tf, TfIdf, BM25, PL2, QL


@fixture
def dataset_name() -> str:
    return "vaswani"


@fixture
def dataset(dataset_name: str) -> Dataset:
    return load(dataset_name)


@fixture
def index_dir(tmp_path: Path, dataset: Dataset) -> Path:
    index_dir = tmp_path
    indexer = IterDictIndexer(
        str(index_dir.absolute()),
        meta={"docno": 20, "text": 5000}
    )
    indexer.index(
        {"docno": doc.doc_id, "text": doc.text}
        for doc in dataset.docs_iter()
    )
    return index_dir


@fixture
def reranking_context(index_dir: Path) -> IndexContext:
    return TerrierIndexContext(index_dir)


@fixture
def context(reranking_context: IndexContext) -> IndexContext:
    return reranking_context


@fixture
def query() -> Query:
    return Query("solving linear equations")


def test_query(query: Query):
    assert query.title == "solving linear equations"


@fixture
def document1(dataset: Dataset) -> Document:
    doc = dataset.docs_iter()[0]
    return Document(doc.doc_id)


def test_document1(document1: Document):
    assert document1.id == "1"


@fixture
def document2(dataset: Dataset) -> Document:
    doc = dataset.docs_iter()[1]
    return Document(doc.doc_id)


def test_document2(document2: Document):
    assert document2.id == "2"


def test_cache_dir(context: IndexContext):
    assert context.cache_dir is None or isinstance(context.cache_dir, Path)


def test_document_count(context: IndexContext):
    assert context.document_count == 11429


def test_document_frequency(context: IndexContext):
    assert context.document_frequency("computer") == 0
    assert context.document_frequency("comput") == 532
    assert context.document_frequency("linear") == 398
    assert context.document_frequency("digital") == 0
    assert context.document_frequency("digit") == 241


def test_inverse_document_frequency(context: IndexContext):
    assert context.inverse_document_frequency("computer") == 0
    assert context.inverse_document_frequency("comput") == approx(3.067266)
    assert context.inverse_document_frequency("linear") == approx(3.357457)
    assert context.inverse_document_frequency("digital") == 0
    assert context.inverse_document_frequency("digit") == approx(3.859112)


def test_contents_query(
        context: IndexContext,
        query: Query
):
    assert context.contents(query) == "solving linear equations"


def test_contents_document1(
        context: IndexContext,
        document1: Document
):
    assert context.contents(document1) == (
        "compact memories have flexible capacities  a digital data "
        "storage\nsystem with capacity up to bits and random and or "
        "sequential access\nis described"
    )


def test_contents_document2(
        context: IndexContext,
        document2: Document
):
    assert context.contents(document2) == (
        "an electronic analogue computer for solving systems of linear "
        "equations\nmathematical derivation of the operating principle "
        "and stability\nconditions for a computer consisting of "
        "amplifiers"
    )


def test_terms_query(context: IndexContext, query: Query):
    assert context.terms(query) == ("solv", "linear", "equat")


def test_terms_document1(context: IndexContext, document1: Document):
    assert context.terms(document1) == (
        "compact", "memori", "flexibl", "capac", "digit", "data", "storag",
        "system", "capac", "bit", "random", "sequenti", "access"
    )


def test_terms_document2(context: IndexContext, document2: Document):
    assert context.terms(document2) == (
        "electron", "analogu", "comput", "solv", "system", "linear",
        "equat", "mathemat", "deriv", "oper", "principl", "stabil",
        "condit", "comput", "consist", "amplifi"
    )


def test_term_set_query(context: IndexContext, query: Query):
    assert context.term_set(query) == {"solv", "linear", "equat"}


def test_term_set_document1(
        context: IndexContext,
        document1: Document
):
    assert context.term_set(document1) == {
        "compact", "memori", "flexibl", "capac", "digit", "data", "storag",
        "system", "bit", "random", "sequenti", "access"
    }


def test_term_set_document2(
        context: IndexContext,
        document2: Document
):
    assert context.term_set(document2) == {
        "electron", "analogu", "comput", "solv", "system", "linear",
        "equat", "mathemat", "deriv", "oper", "principl", "stabil",
        "condit", "comput", "consist", "amplifi"
    }


def test_term_frequency_query(context: IndexContext, query: Query):
    assert context.term_frequency(query, "solv") == approx(1 / 3)
    assert context.term_frequency(query, "linear") == approx(1 / 3)
    assert context.term_frequency(query, "tree") == approx(0 / 3)


def test_term_frequency_document1(
        context: IndexContext,
        document1: Document
):
    assert context.term_frequency(document1, "compact") == approx(1 / 13)
    assert context.term_frequency(document1, "capac") == approx(2 / 13)
    assert context.term_frequency(document1, "tree") == approx(0 / 13)


def test_term_frequency_document2(
        context: IndexContext,
        document2: Document
):
    assert context.term_frequency(document2, "electron") == approx(1 / 16)
    assert context.term_frequency(document2, "analogu") == approx(1 / 16)
    assert context.term_frequency(document2, "tree") == approx(0 / 16)


def test_document1_retrieval_score(
        context: IndexContext,
        query: Query,
        document1: Document
):
    d1 = document1
    assert context.retrieval_score(query, d1, Tf()) == 0
    assert context.retrieval_score(query, d1, TfIdf()) == 0
    assert context.retrieval_score(query, d1, BM25()) == 0
    assert context.retrieval_score(query, d1, PL2()) == 0
    assert context.retrieval_score(query, d1, QL()) == 0


def test_document2_retrieval_score(
        context: IndexContext,
        query: Query,
        document2: Document
):
    d2 = document2
    # FIXME: How can we verify these numbers?
    assert context.retrieval_score(query, d2, Tf()) == 3
    assert context.retrieval_score(query, d2, TfIdf()) == approx(9.823447)
    assert context.retrieval_score(query, d2, BM25()) == approx(17.657991)
    assert context.retrieval_score(query, d2, PL2()) == approx(1.133970)
    assert context.retrieval_score(query, d2, QL()) == approx(1.278038)
