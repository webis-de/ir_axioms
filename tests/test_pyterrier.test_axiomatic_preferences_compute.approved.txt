qid,docno_a,docno_b,_DOC_ID_preferences
q1,doc1,doc1,0.0
q1,doc1,doc3,-1.0
q1,doc1,doc2,-1.0
q1,doc3,doc1,1.0
q1,doc3,doc3,0.0
q1,doc3,doc2,1.0
q1,doc2,doc1,1.0
q1,doc2,doc3,-1.0
q1,doc2,doc2,0.0
q2,doc5,doc5,0.0
q2,doc5,doc6,-1.0
q2,doc5,doc4,1.0
q2,doc6,doc5,1.0
q2,doc6,doc6,0.0
q2,doc6,doc4,1.0
q2,doc4,doc5,-1.0
q2,doc4,doc6,-1.0
q2,doc4,doc4,0.0
